metric/id/m3/id.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package m3 describes m3 metric id information.
+//
+// Each M3 metric id contains a metric name and a set of tag pairs. In particular,
+// it conforms to the following format:
+//
+// m3+<metric_name>+tagName1=tagValue1,tagName2=tagValue2,...tagNameN=tagValueN,
+//
+// Where the tag names are sorted alphabetically in ascending order.
+//
+// An example m3 metrid id is as follows:
+// m3+response_code+env=bar,service=foo,status=404,type=counters
+package m3
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3metrics/metric/id"
+)
+
+const (
+	componentSplitter = '+'
+	tagNameSplitter   = '='
+	tagPairSplitter   = ','
+)
+
+var (
+	errInvalidM3Metric = fmt.Errorf("invalid m3 metric")
+	m3Prefix           = []byte("m3+")
+	nameTagName        = []byte("name")
+	rollupTagPair      = id.TagPair{
+		Name:  []byte("m3_rollup"),
```

Comment:
I know we use this internally, but I'm wondering if it wouldn't be preferable to use a more neutral tag as we transition to open source? I'm thinking something like `__rollup__`. This also has the advantage that since `_` is before the lower case letters in ASCII it might be easier to insert tags with an underscore prefix in the future to the tags pair.